cosi/cosi.go

```diff
@@ -1,450 +1,399 @@
+		}
+		if ((m.mask[byt] & msk) != 0) && ((mask[byt] & msk) == 0) {
+			m.mask[byt] ^= msk // flip bit in mask from 1 to 0
+			m.AggregatePublic.Sub(m.AggregatePublic, m.publics[i])
+		}
+	}
+	return nil
 }
 
-// genCommit generates a random scalar vi and computes its individual commit
-// Vi = G^vi
-func (c *CoSi) genCommit(s cipher.Stream) {
-	var stream = s
-	if s == nil {
-		stream = random.Stream
-	}
-	c.random = c.suite.Scalar().Pick(stream)
-	c.commitment = c.suite.Point().Mul(nil, c.random)
-	c.aggregateCommitment = c.commitment
+// Len returns the mask length in bytes.
+func (m *Mask) Len() int {
+	return (len(m.publics) + 7) >> 3
 }
 
-// genResponse creates the response
-func (c *CoSi) genResponse() error {
-	if c.private == nil {
-		return errors.New("No private key given in this cosi")
+// SetBit enables (enable: true) or disables (enable: false) the bit
+// in the participation mask of the given cosigner.
+func (m *Mask) SetBit(signer int, enable bool) error {
+	if signer > len(m.publics) {
+		return errors.New("index out of range")
 	}
-	if c.random == nil {
-		return errors.New("No random scalar computed in this cosi")
+	byt := signer >> 3
+	msk := byte(1) << uint(signer&7)
+	if ((m.mask[byt] & msk) == 0) && enable {
+		m.mask[byt] ^= msk // flip bit in mask from 0 to 1
+		m.AggregatePublic.Add(m.AggregatePublic, m.publics[signer])
 	}
-	if c.challenge == nil {
-		return errors.New("No challenge computed in this cosi")
+	if ((m.mask[byt] & msk) != 0) && !enable {
+		m.mask[byt] ^= msk // flip bit in mask from 1 to 0
+		m.AggregatePublic.Sub(m.AggregatePublic, m.publics[signer])
 	}
-
-	// resp = random - challenge * privatekey
-	// i.e. ri = vi + c * xi
-	resp := c.suite.Scalar().Mul(c.private, c.challenge)
-	c.response = resp.Add(c.random, resp)
-	// no aggregation here
-	c.aggregateResponse = c.response
-	// paranoid protection: delete the random
-	c.random = nil
 	return nil
 }
 
-// mask holds the mask utilities
-type mask struct {
-	mask      []byte
-	publics   []abstract.Point
-	aggPublic abstract.Point
-	suite     abstract.Suite
+// Bit returns a boolean value indicating whether the given signer is
+// enabled (true) or disabled (false).
+func (m *Mask) Bit(signer int) (bool, error) {
+	if signer > len(m.publics) {
+		return false, errors.New("index out of range")
+	}
+	byt := signer >> 3
+	msk := byte(1) << uint(signer&7)
+	return (m.mask[byt] & msk) != 0, nil
 }
 
-// newMask returns a new mask to use with the cosigning with all cosigners enabled
-func newMask(suite abstract.Suite, publics []abstract.Point) *mask {
-	// Start with an all-disabled participation mask, then set it correctly
-	cm := &mask{
-		publics: publics,
-		suite:   suite,
+// CountEnabled returns the number of enabled nodes in the CoSi participation
+// mask, i.e., it returns the hamming weight of the mask.
+func (m *Mask) CountEnabled() int {
+	hw := 0
+	for i := range m.publics {
+		byt := i >> 3
```

Comment:
Hmm but then I would need to return `(int, error)` in `CountEnabled()` which is not really necessary here I think.